File: nhglib-core/src/io/github/voidzombie/nhglib/runtime/ecs/components/common/MessageComponent.java
package io.github.voidzombie.nhglib.runtime.ecs.components.common;

import com.artemis.PooledComponent;
import com.badlogic.gdx.utils.Array;
import io.github.voidzombie.nhglib.Nhg;
import io.github.voidzombie.nhglib.runtime.messaging.Message;
import io.reactivex.functions.Consumer;

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
public class MessageComponent extends PooledComponent {
    private Array<Message> messages;
    private Array<String> filters;

    public MessageComponent() {
        messages = new Array<>();
        filters = new Array<>();
    }

    @Override
    protected void reset() {
        messages.clear();
        filters.clear();
    }

    public void subscribe(String ... filters) {
        Nhg.messaging.get(filters)
                .subscribe(new Consumer<Message>() {
                    @Override
                    public void accept(Message message) throws Exception {
                        messages.add(message);
                    }
                });

        this.filters.clear();
        this.filters.addAll(filters);
    }

    public void consume(Message message) {
        messages.removeValue(message, false);
    }

    public Array<Message> getMessages() {
        return messages;
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/core/ecs/components/common/MessageComponent.kt
package io.github.movementspeed.nhglib.core.ecs.components.common

import com.artemis.PooledComponent
import com.badlogic.gdx.utils.Array
import io.github.movementspeed.nhglib.core.messaging.Message

/**
 * Created by Fausto Napoli on 08/12/2016.
 */
class MessageComponent : PooledComponent() {
    val messages: Array<Message>
    private val filters: Array<String>

    init {
        messages = Array()
        filters = Array()
    }

    override fun reset() {
        messages.clear()
        filters.clear()
    }

    fun subscribe(vararg filters: String) {
        /*Nhg.messaging.get(filters)
                .subscribe(new Consumer<Message>() {
                    @Override
                    public void accept(Message message) throws Exception {
                        messages.add(message);
                    }
                });*/

        this.filters.clear()
        this.filters.addAll(*filters)
    }

    fun consume(message: Message) {
        messages.removeValue(message, false)
    }
}
